File: python/evidence_gate/verify.py
"""Evidence Gate — post-generation claim verification (Python).

The gate decides WHETHER the model may speak; verify_claims checks whether
what it said stands on the evidence it was given. The core cannot judge
semantic truth — it verifies a citation protocol, deterministically and
identically across the JS and Python ports:

1. citation_block() renders the markers the model must cite from.
2. verify_claims() parses the answer: every citation must resolve to a
   record that exists (no phantom evidence), every claim-looking sentence
   must carry a citation (no naked claims), and the framing must match the
   gate's verdict (no "as of today" over stale data).

Spec: docs/design/claim-verification.md (approved). No dependencies.
"""
import re
from datetime import datetime, timezone

from .core import evidence_digest, canonical_json

VERIFICATION_SCHEMA = "evidence-gate.verification/1"

# Marker grammar, shared verbatim with the JS port — one grammar only.
_MARKER_RE = re.compile(r"\[ev:([A-Za-z0-9_.-]+)\]")
# Default claim patterns: any digit, or a quantity-implying symbol.
# Regex SOURCES (strings), never functions — rulesets stay JSON-digestable.
_DEFAULT_CLAIM_PATTERNS = ["\\d", "%|\\$|€|£|฿"]
# English-only by default; apps localize via rules["verification"]["freshness_patterns"].
_DEFAULT_FRESHNESS_PATTERNS = ["\\b(today|as of now|currently|latest|real[- ]?time)\\b"]

_ID_RE = re.compile(r"^[A-Za-z0-9_.-]+$")
_NUMERIC_RE = re.compile(r"^[0-9]+$")
# Explicit ASCII whitespace classes on purpose: JS \s and Python \s disagree
# on exotic Unicode whitespace, and the sentence partition must be
# byte-identical across ports.
_LINE_SPLIT = re.compile(r"\r?\n")
_SENTENCE_SPLIT = re.compile(r"(?<=[.!?])[ \t]+")
_STRIP_CHARS = " \t\r\f\v"


def _usable_id(record_id):
    """A usable record id must fit the marker grammar and must NOT be purely
    numeric — numeric refs are reserved for index resolution, so a numeric id
    could never be told apart from an index. Invalid ids are ignored (the
    record falls back to its index marker), never an error."""
    return (isinstance(record_id, str) and _ID_RE.match(record_id) is not None
            and _NUMERIC_RE.match(record_id) is None)


def citation_block(records=None, opts=None):
    """Render the evidence list the model must cite from.

    opts: {"header"?, "line"?: callable(record, marker, index) -> str, "supporting"?}
    Markers are the record's id when usable, else the 1-based index into the
    concatenation [records..., supporting...]. Supporting records are tagged.
    """
    records = records or []
    opts = opts or {}
    supporting = opts.get("supporting") or []
    header = opts.get("header")
    if header is None:
        header = "EVIDENCE (cite with its marker after every factual statement):"
    lines = [header]
    entries = [(r, "primary") for r in records] + [(r, "supporting") for r in supporting]
    for i, (r, tier) in enumerate(entries):
        r = r or {}
        marker = r.get("id") if _usable_id(r.get("id")) else str(i + 1)
        line_fn = opts.get("line")
        if callable(line_fn):
            lines.append(line_fn(r, marker, i))
            continue
        line = f'[ev:{marker}] {r.get("date") or "undated"}'
        q = r.get("quality_score")
        if isinstance(q, (int, float)) and not isinstance(q, bool):
            line += f" — quality {canonical_json(q)}"
        wf = r.get("flags")
        flags = wf if isinstance(wf, list) else ([wf] if isinstance(wf, str) and wf else [])
        if flags:
            line += f', flags: {", ".join(flags)}'
        if tier == "supporting":
            line += " (supporting)"
        lines.append(line)
    return "\n".join(lines)


def _resolve_ref(ref, records, supporting):
    """ref -> 0-based index into [records..., supporting...] or None (phantom).

    Order: exact id match (records first, then supporting) -> else a numeric
    ref 1..N is that 1-based position."""
    for i, r in enumerate(records):
        if r and _usable_id(r.get("id")) and r.get("id") == ref:
            return {"record": i, "tier": "primary"}
    for j, r in enumerate(supporting):
        if r and _usable_id(r.get("id")) and r.get("id") == ref:
            return {"record": len(records) + j, "tier": "supporting"}
    if _NUMERIC_RE.match(ref):
        n = int(ref)
        if 1 <= n <= len(records) + len(supporting):
            return {"record": n - 1, "tier": "primary" if n <= len(records) else "supporting"}
    return None


def _split_sentences(text):
    """A stable, port-identical partition: split on newlines, then after
    [.!?] followed by a space/tab."""
    out = []
    for line in _LINE_SPLIT.split(text):
        for part in _SENTENCE_SPLIT.split(line):
            s = part.strip(_STRIP_CHARS)
            if s:
                out.append(s)
    return out


def verify_claims(answer=None, records=None, supporting=None, gate=None, rules=None, decision=None):
    """The post-generation half of the proof loop.

    -> {pass, verdict, citations, claims, stats, warnings, caveats, decision?}

    Never raises on messy input — a garbage answer yields a verdict, not an
    exception. ``rules`` is optional here (only "verification" and "messages"
    are read); pass the SAME records/supporting the prompt was built from, in
    the same order, or resolution and the digest join are meaningless.
    """
    answer = "" if answer is None else str(answer)
    records = records or []
    supporting = supporting or []
    v = (rules or {}).get("verification") or {}
    m = (rules or {}).get("messages") or {}
    rfc = v.get("require_full_coverage")
    require_full_coverage = True if rfc is None else bool(rfc)
    # re.ASCII so \d, \b, \w mean the same thing as the JS port's regexes
    claim_res = [re.compile(s, re.ASCII) for s in (v.get("claim_patterns") or _DEFAULT_CLAIM_PATTERNS)]
    fresh_res = [re.compile(s, re.ASCII | re.IGNORECASE)
                 for s in (v.get("freshness_patterns") or _DEFAULT_FRESHNESS_PATTERNS)]

    # every marker occurrence, in order of appearance
    citations = []
    for match in _MARKER_RE.finditer(answer):
        ref = match.group(1)
        resolved = _resolve_ref(ref, records, supporting)
        citations.append({
            "marker": "ev:" + ref,
            "ref": ref,
            "record": resolved["record"] if resolved else None,
            "valid": resolved is not None,
            "tier": resolved["tier"] if resolved else None,
        })

    # claim-looking sentences; markers are stripped before pattern matching so
    # a digit inside [ev:1] can never make a sentence look like a claim
    claims = []
    cited = 0
    uncited = 0
    for sentence in _split_sentences(answer):
        refs = _MARKER_RE.findall(sentence)
        stripped = _MARKER_RE.sub("", sentence)
        if not any(p.search(stripped) for p in claim_res):
            continue
        has_valid = any(_resolve_ref(ref, records, supporting) is not None for ref in refs)
        claims.append({"text": sentence, "cited": has_valid, "markers": ["ev:" + r for r in refs]})
        if has_valid:
            cited += 1
        else:
            uncited += 1

    phantom = sum(1 for c in citations if not c["valid"])
    valid_count = len(citations) - phantom
    stats = {"claims": len(claims), "cited": cited, "uncited": uncited, "phantom": phantom}

    # verdict ladder — top-down, first hit wins
    if phantom > 0:
        verdict, passed = "phantom_citations", False
    elif claims and valid_count == 0:
        verdict, passed = "no_citations", False
    elif uncited > 0:
        verdict, passed = "unsupported_claims", not require_full_coverage
    else:
        verdict, passed = "supported", True  # no claims at all (a refusal) passes

    warnings = []
    if phantom > 0:
        seen, listed = set(), []
        for c in citations:
            if not c["valid"] and c["marker"] not in seen:
                seen.add(c["marker"])
                listed.append(f'[{c["marker"]}]')
        warnings.append({"level": "block", "code": "verify_phantom_citation",
                         "message": m.get("verify_phantom_citation",
                                          f'Citations resolve to no evidence record: {", ".join(listed)} — the model may have invented sources.')})
    if claims and valid_count == 0:
        warnings.append({"level": "block", "code": "verify_no_citations",
                         "message": m.get("verify_no_citations",
                                          "The answer makes claims but cites no evidence — the citation protocol was ignored.")})
    elif uncited > 0:
        warnings.append({"level": "review", "code": "verify_uncited_claim",
                         "message": m.get("verify_uncited_claim",
                                          f"{uncited} claim(s) lack a valid citation — treat them as unverified.")})
    if gate and gate.get("freshness") == "stale" and any(p.search(answer) for p in fresh_res):
        warnings.append({"level": "review", "code": "verify_stale_framing",
                         "message": m.get("verify_stale_framing",
                                          'The answer uses fresh-sounding framing but the evidence is stale — the AI must not say "latest" or "today".')})

    result = {"pass": passed, "verdict": verdict, "citations": citations, "claims": claims,
              "stats": stats, "warnings": warnings, "caveats": [w["message"] for w in warnings]}

    if decision:
        meta = {} if decision is True else decision
        at = meta.get("at")
        if at is None:
            at = datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
        result["decision"] = {
            "schema": VERIFICATION_SCHEMA,
            "id": meta.get("id"),
            "at": at,
            # evidence digest: identical function + input as the gate's decision
            # record -> the join key. Neither the evidence nor the answer is stored.
            "digests": {"evidence": evidence_digest({"records": records, "supporting": supporting}),
                        "answer": evidence_digest(answer)},
            "verdict": verdict,
            "pass": passed,
            "stats": stats,
            "warnings": [{"level": w["level"], "code": w["code"]} for w in warnings],
        }

    return result

File: src/verify.js
// Evidence Gate — post-generation claim verification
//
// The gate decides WHETHER the model may speak; verifyClaims checks whether
// what it said stands on the evidence it was given. The core cannot judge
// semantic truth — it verifies a citation protocol, deterministically and
// identically across the JS and Python ports:
//
//   1. citationBlock() renders the markers the model must cite from.
//   2. verifyClaims() parses the answer: every citation must resolve to a
//      record that exists (no phantom evidence), every claim-looking sentence
//      must carry a citation (no naked claims), the framing must match the
//      gate's verdict (no "as of today" over stale data), and — when records
//      carry facts — every cited number must match them (no misquoted values).
//
// Spec: docs/design/claim-verification.md (approved). Zero dependencies.

import { evidenceDigest, canonicalJson } from "./core.js";

export const VERIFICATION_SCHEMA = "evidence-gate.verification/1";

// Marker grammar, shared verbatim with the Python port — one grammar only.
const MARKER_RE_SOURCE = "\\[ev:([A-Za-z0-9_.-]+)\\]";
// Default claim patterns: any digit (ASCII or Thai — the Thai range is
// explicit, never Unicode \d), or a quantity-implying symbol.
// Regex SOURCES (strings), never functions — rulesets stay JSON-digestable.
const DEFAULT_CLAIM_PATTERNS = ["\\d", "[๐-๙]", "%|\\$|€|£|฿"];
// English-only by default; apps localize via rules.verification.freshnessPatterns.
const DEFAULT_FRESHNESS_PATTERNS = ["\\b(today|as of now|currently|latest|real[- ]?time)\\b"];

const ID_RE = /^[A-Za-z0-9_.-]+$/;
const NUMERIC_RE = /^[0-9]+$/;

// A usable record id must fit the marker grammar and must NOT be purely
// numeric — numeric refs are reserved for index resolution, so a numeric id
// could never be told apart from an index. Invalid ids are ignored (the
// record falls back to its index marker), never an error.
function usableId(id) {
  return typeof id === "string" && ID_RE.test(id) && !NUMERIC_RE.test(id);
}

// ── citationBlock: render the evidence list the model must cite from ─────────
//   citationBlock(records, opts?) → string
//   opts: { header?, line?: (record, marker, index) => string, supporting? }
//   Markers are the record's id when usable, else the 1-based index into the
//   concatenation [records..., supporting...]. Supporting records are tagged.
export function citationBlock(records = [], opts = {}) {
  records = records || [];
  const supporting = opts.supporting || [];
  const header = opts.header ?? "EVIDENCE (cite with its marker after every factual statement):";
  const lines = [header];
  const all = [...records.map((r) => [r, "primary"]), ...supporting.map((r) => [r, "supporting"])];
  all.forEach(([r, tier], i) => {
    r = r || {};
    const marker = usableId(r.id) ? r.id : String(i + 1);
    if (typeof opts.line === "function") { lines.push(opts.line(r, marker, i)); return; }
    let line = `[ev:${marker}] ${r.date || "undated"}`;
    if (typeof r.qualityScore === "number") line += ` — quality ${canonicalJson(r.qualityScore)}`;
    const flags = Array.isArray(r.flags) ? r.flags : typeof r.flags === "string" && r.flags ? [r.flags] : [];
    if (flags.length) line += `, flags: ${flags.join(", ")}`;
    if (tier === "supporting") line += " (supporting)";
    lines.push(line);
  });
  return lines.join("\n");
}

// ── resolution: ref → 0-based index into [records..., supporting...] ─────────
// Order: exact id match (records first, then supporting) → else a numeric ref
// 1..N is that 1-based position. Anything else is phantom (null).
function resolveRef(ref, records, supporting) {
  for (let i = 0; i < records.length; i++) {
    const r = records[i];
    if (r && usableId(r.id) && r.id === ref) return { record: i, tier: "primary" };
  }
  for (let j = 0; j < supporting.length; j++) {
    const r = supporting[j];
    if (r && usableId(r.id) && r.id === ref) return { record: records.length + j, tier: "supporting" };
  }
  if (NUMERIC_RE.test(ref)) {
    const n = parseInt(ref, 10);
    if (n >= 1 && n <= records.length + supporting.length)
      return { record: n - 1, tier: n <= records.length ? "primary" : "supporting" };
  }
  return null;
}

// ── fact cross-checking: deterministic misquote detection (spec §8) ───────────
// Extraction pipeline and matching rule are normative in the design doc and
// must stay byte-identical across ports: Thai digit translation → ISO date
// masking → one shared token regex → decimal-point-shift canonicalization.

// ISO dates are masked so date components never leak into strict matching.
const ISO_DATE_RE_SOURCE = "(?<![0-9])[0-9]{4}-[0-9]{2}-[0-9]{2}(?![0-9])";
// Groups: 1 = sign, 2 = integer part (comma separators), 3 = .fraction, 4 = suffix.
// The lookbehind keeps ranges positive (3-5% → 3, 5) and stops v1.2.3 tails;
// the lookahead keeps 512MB / 10Kg from reading as magnitudes.
const NUMBER_TOKEN_RE_SOURCE =
  "(?<![0-9.])(-?)([0-9]{1,3}(?:,[0-9]{3})+|[0-9]+)((?:\\.[0-9]+)?)" +
  "(?:[ \\t]?(ล้านล้าน|แสนล้าน|หมื่นล้าน|พันล้าน|ล้าน|แสน|หมื่น|พัน|[KkMmBb](?![A-Za-z0-9])))?";
// Fixed magnitude table (suffix → power of ten), NOT overridable.
const MAGNITUDES = {
  K: 3, k: 3, M: 6, m: 6, B: 9, b: 9,
  "พัน": 3, "หมื่น": 4, "แสน": 5, "ล้าน": 6,
  "พันล้าน": 9, "หมื่นล้าน": 10, "แสนล้าน": 11, "ล้านล้าน": 12,
};

// Thai numerals ๐-๙ (U+0E50–U+0E59) → 0-9 via an explicit offset, never
// Unicode \d (the Python port compiles every pattern with re.ASCII).
function translateThaiDigits(text) {
  return text.replace(/[๐-๙]/g, (c) => String.fromCharCode(c.charCodeAt(0) - 0x0e50 + 48));
}

// Canonical value: shift the decimal point in the digit string, then parse
// ONCE. Never multiply by a float power of ten — 1.2 * 1e6 is NOT 1200000 in
// IEEE-754, and exact matching would break identically in both ports.
function tokenValue(sign, intPart, fracPart, suffix) {
  const I = intPart.replace(/,/g, "");
  const F = fracPart ? fracPart.slice(1) : "";
  const e = suffix ? MAGNITUDES[suffix] : 0;
  const digits = I + F;
  const p = I.length + e;
  let whole, frac;
  if (p >= digits.length) { whole = digits + "0".repeat(p - digits.length); frac = ""; }
  else { whole = digits.slice(0, p); frac = digits.slice(p); }
  whole = whole.replace(/^0+/, "") || "0";
  frac = frac.replace(/0+$/, "");
  return parseFloat(sign + whole + (frac ? "." + frac : ""));
}

// Usable facts = a plain object with ≥1 finite numeric value (booleans are
// not numbers; {} is NOT usable — spelled out to dodge the truthiness trap).
function factValues(record) {
  const f = record ? record.facts : null;
  if (!f || typeof f !== "object" || Array.isArray(f)) return [];
  const out = [];
  for (const k of Object.keys(f)) {
    if (typeof f[k] === "number" && Number.isFinite(f[k])) out.push(f[k]);
  }
  return out;
}

// ── sentence segmentation: a stable, port-identical partition ─────────────────
// Split on newlines, then after [.!?] followed by a space/tab. Explicit ASCII
// whitespace classes on purpose: JS \s and Python \s disagree on exotic
// Unicode whitespace, and this partition must be byte-identical across ports.
function splitSentences(text) {
  const out = [];
  for (const line of text.split(/\r?\n/)) {
    for (const part of line.split(/(?<=[.!?])[ \t]+/)) {
      const s = part.replace(/^[ \t\r\f\v]+|[ \t\r\f\v]+$/g, "");
      if (s) out.push(s);
    }
  }
  return out;
}

// ── verifyClaims: the post-generation half of the proof loop ─────────────────
//   verifyClaims({ answer, records, supporting?, gate?, rules?, decision? })
//     → { pass, verdict, citations, claims, stats, warnings, caveats, decision? }
//   Never throws on messy input — a garbage answer yields a verdict, not an
//   exception. `rules` is optional here (only .verification and .messages are
//   read); pass the SAME records/supporting the prompt was built from, in the
//   same order, or resolution and the digest join are meaningless.
export function verifyClaims({ answer, records = [], supporting = [], gate, rules, decision } = {}) {
  answer = answer == null ? "" : String(answer);
  records = records || [];
  supporting = supporting || [];
  const V = (rules && rules.verification) || {};
  const M = (rules && rules.messages) || {};
  const requireFullCoverage = V.requireFullCoverage == null ? true : !!V.requireFullCoverage;
  // Pattern sources are compiled per port. The Python port compiles with
  // re.ASCII so \d, \b, \w mean the same thing in both languages.
  const claimRes = (V.claimPatterns || DEFAULT_CLAIM_PATTERNS).map((s) => new RegExp(s));
  const freshRes = (V.freshnessPatterns || DEFAULT_FRESHNESS_PATTERNS).map((s) => new RegExp(s, "i"));

  // every marker occurrence, in order of appearance
  const citations = [];
  for (const m of answer.matchAll(new RegExp(MARKER_RE_SOURCE, "g"))) {
    const ref = m[1];
    const resolved = resolveRef(ref, records, supporting);
    citations.push({
      marker: "ev:" + ref,
      ref,
      record: resolved ? resolved.record : null,
      valid: resolved !== null,
      tier: resolved ? resolved.tier : null,
    });
  }

  // claim-looking sentences; markers are stripped before pattern matching so
  // a digit inside [ev:1] can never make a sentence look like a claim
  const claims = [];
  const misquotes = [];
  let cited = 0, uncited = 0;
  for (const sentence of splitSentences(answer)) {
    const refs = [...sentence.matchAll(new RegExp(MARKER_RE_SOURCE, "g"))].map((m) => m[1]);
    const stripped = sentence.replace(new RegExp(MARKER_RE_SOURCE, "g"), "");
    if (!claimRes.some((re) => re.test(stripped))) continue;
    const hasValid = refs.some((ref) => resolveRef(ref, records, supporting) !== null);
    claims.push({ text: sentence, cited: hasValid, markers: refs.map((r) => "ev:" + r) });
    hasValid ? cited++ : uncited++;

    // fact cross-check (§8): pool = union of numeric facts of every validly
    // cited record; empty pool = opt-out, the sentence is skipped silently.
    if (!hasValid) continue;
    const pool = [];
    for (const ref of refs) {
      const resolved = resolveRef(ref, records, supporting);
      if (!resolved) continue;
      const rec = resolved.record < records.length
        ? records[resolved.record] : supporting[resolved.record - records.length];
      pool.push(...factValues(rec));
    }
    if (pool.length === 0) continue;
    const masked = translateThaiDigits(stripped).replace(new RegExp(ISO_DATE_RE_SOURCE, "g"), " ");
    for (const m of masked.matchAll(new RegExp(NUMBER_TOKEN_RE_SOURCE, "g"))) {
      const value = tokenValue(m[1], m[2], m[3], m[4]);
      if (!pool.includes(value)) misquotes.push({ token: m[0], value, sentence });
    }
  }

  const phantom = citations.filter((c) => !c.valid).length;
  const validCount = citations.length - phantom;
  const stats = { claims: claims.length, cited, uncited, phantom, misquoted: misquotes.length };

  // verdict ladder — top-down, first hit wins
  let verdict, pass;
  if (phantom > 0) { verdict = "phantom_citations"; pass = false; }
  else if (misquotes.length > 0) { verdict = "misquoted_values"; pass = false; }
  else if (claims.length > 0 && validCount === 0) { verdict = "no_citations"; pass = false; }
  else if (uncited > 0) { verdict = "unsupported_claims"; pass = !requireFullCoverage; }
  else { verdict = "supported"; pass = true; } // no claims at all (a refusal) passes

  const warnings = [];
  if (phantom > 0) {
    const seen = new Set(), list = [];
    for (const c of citations) if (!c.valid && !seen.has(c.marker)) { seen.add(c.marker); list.push(`[${c.marker}]`); }
    warnings.push({ level: "block", code: "verify_phantom_citation",
      message: M.verify_phantom_citation || `Citations resolve to no evidence record: ${list.join(", ")} — the model may have invented sources.` });
  }
  if (misquotes.length > 0) {
    const seen = new Set(), list = [];
    for (const q of misquotes) if (!seen.has(q.token)) { seen.add(q.token); list.push(q.token); }
    warnings.push({ level: "block", code: "verify_misquoted_value",
      message: M.verify_misquoted_value || `Cited numbers do not match the evidence facts: ${list.join(", ")} — the model may have misquoted values.` });
  }
  if (claims.length > 0 && validCount === 0)
    warnings.push({ level: "block", code: "verify_no_citations",
      message: M.verify_no_citations || "The answer makes claims but cites no evidence — the citation protocol was ignored." });
  else if (uncited > 0)
    warnings.push({ level: "review", code: "verify_uncited_claim",
      message: M.verify_uncited_claim || `${uncited} claim(s) lack a valid citation — treat them as unverified.` });
  if (gate && gate.freshness === "stale" && freshRes.some((re) => re.test(answer)))
    warnings.push({ level: "review", code: "verify_stale_framing",
      message: M.verify_stale_framing || `The answer uses fresh-sounding framing but the evidence is stale — the AI must not say "latest" or "today".` });

  const result = { pass, verdict, citations, claims, misquotes, stats, warnings, caveats: warnings.map((w) => w.message) };

  if (decision) {
    const meta = decision === true ? {} : decision;
    result.decision = {
      schema: VERIFICATION_SCHEMA,
      id: meta.id ?? null,
      at: meta.at ?? new Date().toISOString(),
      // evidence digest: identical function + input as the gate's decision
      // record → the join key. Neither the evidence nor the answer is stored.
      digests: { evidence: evidenceDigest({ records, supporting }), answer: evidenceDigest(answer) },
      verdict,
      pass,
      stats,
      warnings: warnings.map(({ level, code }) => ({ level, code })),
    };
  }

  return result;
}
